### DISCUSSION TITLE
GPIO Interrupt

### DISCUSSION THREAD

**Discussion by sheershak-agarwal** [nasa/fprime#1009] [Q&A]
Hi I want to integrate gps pps into my system and i want the exact time when my pps goes high through the gpio. I saw the setting GPIO_INT. How do I use it?

**Comment by timcanham** [comment]
What GPIO_INT are you referring to?

**Reply by sheershak-agarwal** [reply]
https://github.com/nasa/fprime/blob/0845912142ec08c5fbee320a6d55da94a993c544/Drv/LinuxGpioDriver/LinuxGpioDriverComponentImpl.hpp#L54

**Reply by timcanham** [reply]
Ah - if you configure it as `GPIO_INT`, it will configure the GPIO to invoke the `intOut` port. The port contains an argument with the latched time of the interrupt. For Linux, it is system time.

**Reply by sheershak-agarwal** [reply]
Okay, so it won't really help me get pps time. How do you recommend getting pps signal?

**Comment by timcanham** [comment]
If you tie the GPS PPS to a GPIO, then the time in the port call would be the system time of the pulse. If the pulse is a "time at the tone," you could use that to correlate system time to GPS time.

**Reply by sheershak-agarwal** [reply]
by port call do you mean when I read the gpio? How can we make sure that we are reading the gpio when it is high?

**Reply by timcanham** [reply]
It looks like the interrupt is hard-coded for a rising edge:

https://github.com/nasa/fprime/blob/0845912142ec08c5fbee320a6d55da94a993c544/Drv/LinuxGpioDriver/LinuxGpioDriverComponentImpl.cpp#L333

So if you set the direction as `GPIO_INT`, the driver will call the `intOut` output port whenever the driver sees the GPIO go high. You don't have to call anything, but you have to start the task waiting on the interrupt:

https://github.com/nasa/fprime/blob/0845912142ec08c5fbee320a6d55da94a993c544/Drv/LinuxGpioDriver/LinuxGpioDriverComponentImpl.hpp#L48

You would add a `Svc::Cycle` port to your component that wants to receive the interrupt and connect it to that port on the GPIO driver. The GPIO driver is modal; you can poll it for a change or you can configure it to interrupt on change.

**Reply by sheershak-agarwal** [reply]
Thank you. I will try that tomorrow and ask more question if stuck

**Reply by sheershak-agarwal** [reply]
Ok so what I am doing is:
1. Add the Svc::Cycle port in my Gps module
2. Connect the Svc::Cycle input port with the GPIO Port in the AppTopologyAi.xml
3. Generate new cpp and hpp file and then use the cycle in function for timing correction.

**Comment by sheershak-agarwal** [comment]
Hello, so I am doing the sets below and facing an error.

1. Add PpsRead Port in my Gps Component.
`<port name="PpsRead" data_type="Svc::Cycle" kind="async_input"  max_number="1">
            <comment>
            Output Port for reading GPIO values
            </comment>
        </port>`
2. Create new Impl.cpp and hpp files
3. Open the gpio26Driver (assigned to pps) as GPIO_INT in Topology.cpp
`
if (not gpio26Drv.open(26,Drv::LinuxGpioDriverComponentImpl::GPIO_INT)) {
      Fw::Logger::logMsg("[ERROR] Failed to open PS GPIO 26\n");
    } else {
      Fw::Logger::logMsg("[INFO] Opened PS GPIO 26\n");
    }
`
4. Connected the ppsRead port to the gpioDriver
`<connection name = "gpio26">
       <source component = "gps" port = "PpsRead" type = "Svc::Cycle" num = "0"/>
        <target component = "gpio26Drv" port = "inOut" type = "Svc::Cycle" num = "0"/>
   </connection>`

During the fprime-util install, I got this error:
`
/home/sheershak/Desktop/F-Prime-Debug/App/build-fprime-automatic-raspberrypi/App/Top/AppTopologyAppAc.cpp:292:9: error: ‘class App::GpsComponentImpl’ has no member named ‘set_PpsRead_OutputPort’; did you mean ‘get_PpsRead_InputPort’?
     gps.set_PpsRead_OutputPort(0, gpio26Drv.get_inOut_InputPort(0));
         ^~~~~~~~~~~~~~~~~~~~~~
         get_PpsRead_InputPort
/home/sheershak/Desktop/F-Prime-Debug/App/build-fprime-automatic-raspberrypi/App/Top/AppTopologyAppAc.cpp:292:45: error: ‘class Drv::LinuxGpioDriverComponentImpl’ has no member named ‘get_inOut_InputPort’; did you mean ‘set_intOut_OutputPort’?
App/Top/CMakeFiles/App.dir/build.make:148: recipe for target 'App/Top/CMakeFiles/App.dir/AppTopologyAppAc.cpp.obj' failed
     gps.set_PpsRead_OutputPort(0, gpio26Drv.get_inOut_InputPort(0));
                                             ^~~~~~~~~~~~~~~~~~~
                                             set_intOut_OutputPort
make[2]: *** [App/Top/CMakeFiles/App.dir/AppTopologyAppAc.cpp.obj] Error 1
CMakeFiles/Makefile2:6473: recipe for target 'App/Top/CMakeFiles/App.dir/all' failed
make[1]: *** [App/Top/CMakeFiles/App.dir/all] Error 2
Makefile:135: recipe for target 'all' failed
make: *** [all] Error 2
[ERROR] CMake erred with return code 2
`

**Reply by sheershak-agarwal** [reply]
So the question here is, in my gps component, should the kind be async_input or output? and where do I  start the task waiting on the interrupt?

**Comment by timcanham** [comment]
The GPIO from the GPS receiver is driving your component, so the GPIO driver should be the source. Your GPS component would be `async_input` or `sync_input`. `async_input` would dispatch a message, while `sync_input` would invoke your handler directly, so you may have to guard your data against multiple threads.

**Reply by timcanham** [reply]
As for a task, you could make your GPS component active, so it would have it's own thread to service the interrupt, since it is called by the GPIO driver.

**Reply by sheershak-agarwal** [reply]
The GPS Component is active. Oh ok I will make the gpioDriver as source and the cycle as target. Thanks
